src/main/java/com/google/devtools/build/lib/runtime/commands/TargetPatternFileSupport.java

```diff
@@ -0,0 +1,63 @@
+package com.google.devtools.build.lib.runtime.commands;
+
+import com.google.common.collect.Lists;
+import com.google.devtools.build.lib.buildtool.BuildRequestOptions;
+import com.google.devtools.build.lib.profiler.Profiler;
+import com.google.devtools.build.lib.profiler.SilentCloseable;
+import com.google.devtools.build.lib.runtime.BlazeCommand;
+import com.google.devtools.build.lib.runtime.CommandEnvironment;
+import com.google.devtools.build.lib.vfs.FileSystemUtils;
+import com.google.devtools.build.lib.vfs.Path;
+import com.google.devtools.common.options.OptionsParsingResult;
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.util.List;
+
+/**
+ * Provides support for implementations for {@link BlazeCommand} to read target patterns from file.
+ */
+final class TargetPatternFileSupport {
+
+  private TargetPatternFileSupport() {}
+
+  /**
+   * Reads a list of target patterns, either from the command-line residue or by reading newline
+   * delimited target patterns from the --target_pattern_file flag. If --target_pattern_file is
+   * specified and options contain a residue, or file cannot be read it throws an exception instead.
+   */
+  public static List<String> handleTargetPatternFile(
```

Comment:
nit: this isn't always getting the target patterns from a file, in fact it's almost always going to come from the commandline. Can we rename this to better reflect that? I think renaming the class/method to something like "TargetPatternsHelper#readFrom" avoids committing to too much.